### ISSUE TITLE
InjectionToken not resolved in downgraded component

### ISSUE THREAD

**Issue by aciccarello** [angular/angular#19729]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>
Note: I originally [posted this](https://stackoverflow.com/q/46738731/4252741) to Stack Overflow but now suspect that this is a bug in the dependency injection setup for hybrid apps

## Current behavior

Downgraded components do not appear in the page because of errors when initially loading a Hybrid Angular/AngularJS route. In below example a downgraded Angular component fails to load because the dependency injection system does not resolve an `InjectionToken` the material menu needs. The method of bootstrapping AngularJS in an Angular app is [based on Victor Savkin work][2].

```
TypeError: Cannot read property 'xPosition' of undefined
    at new MatMenu (VM4642 material.umd.js:13096)
    at createClass (VM4479 core.umd.js:10963)
    at createDirectiveInstance (VM4479 core.umd.js:10788)
    at createViewNodes (VM4479 core.umd.js:12229)
    at callViewAction (VM4479 core.umd.js:12675)
    at execComponentViewsAction (VM4479 core.umd.js:12584)
    at createViewNodes (VM4479 core.umd.js:12256)
    at createRootView (VM4479 core.umd.js:12124)
    at callWithDebugContext (VM4479 core.umd.js:13507)
    at Object.debugCreateRootView [as createRootView] (VM4479 core.umd.js:12824) "<my-layout class="ng-scope" ng-version="4.3.6">"
```

By stepping through the code I can see that object that `xPosition` is on is an [injected default options object][3]. The `MatMenu` depends on this default options object that is referenced via an InjectionToken. This dependency is coming back undefined when reloading the page. However in my hybrid Angular/AngularJS application, the dependency loads correctly if switching to that page from another page.

Below is a simplified view of the dom structure. Note that ng1 and ng2 components are interlaced. The above error refers to the top downgraded component rather than the downgraded component that wraps the `<md-menu>`.

```html
<ng2-root>
  <router-outlet></router-outlet>
  <div ui-view>
    <downgraded-ng2-component>
      <ng1-component>
        <another-downgraded-ng2-component>
          <button md-icon-button [mdMenuTriggerFor]"toolbarMenu"></button>
          <md-menu #toolbarMenu="mdMenu">
            <button md-menu-item>...</button>
          </md-menu>
        </another-downgraded-ng2-component>
      </ng1-component>
    </downgraded-ng2-component>
  </div>
</ng2-root>
```

  [1]: https://stackblitz.com/edit/ajc-angular-upgrade-di-error?embed=1&file=app/layout.component.html
  [2]: https://blog.nrwl.io/using-ngupgrade-like-a-pro-lazy-loading-angularjs-applications-469819f5c86
  [3]: https://github.com/angular/material2/blob/9673f63c11a286639e5d3a3b9e379392fd666d2a/src/lib/menu/menu-directive.ts#L150

## Expected behavior

The dependency injection should work for the downgraded component such that the dropdown can load without issue.

## Minimal reproduction of the problem with instructions

1. Open [this stackblitz app](https://stackblitz.com/edit/ajc-angular-upgrade-di-error?embed=1&file=app/layout.component.html)
2. Click the link in the app to load an AngularJS page
3. Reload the app iFrame (on the angular JS route)
4. The button will not load and the console will show the error `Cannot read property 'xPosition' of undefined`

## What is the motivation / use case for changing the behavior?

Dependency injection should work for downgraded Angular components

## Environment

<pre><code>
Angular version: 4.3.6

Browser:
- [x ] Chrome (desktop) version 61.0.3163.100 (Official Build) (64-bit)
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

Others:

@angular/material: 2.0.0-beta.12
</code></pre>

**Comment by gkalpak** [comment]
**tl;dr**
- You need to bootstrap AngularJS asynchronously, to give time to all providers to be initialized.
- Therefore, (synchronously) bootstrapping AngularJS from the constructor of a module is probably not the best idea, but you should be able to work-around that easily. 

I have left a [more detailed answer][1] on StackOverflow.

##
Closing since everything is working as expected (afaict) and synchronously bootstrapping in a module's constructor is not something we "officially recommend" or document as an option.

(Just to be clear, I am not saying lazy-loading AngularJS should be avoided or anything, just make sure you do it at a point in time where all Angular parts are guaranteed to be in place.)

[1]: https://stackoverflow.com/questions/46738731/why-isnt-arent-md-menu-options-injected-in-my-hybrid-angular-application/46768876#answer-46768876
